apollo-client/src/main/java/com/ctrip/framework/apollo/internals/AbstractConfig.java

```diff
@@ -450,7 +468,7 @@ public void run() {
           String listenerName = listener.getClass().getName();
           Transaction transaction = Tracer.newTransaction("Apollo.ConfigChangeListener", listenerName);
           try {
-            listener.onChange(changeEvent);
+            listener.onChange(resolve(listener, changeEvent));
```

Comment:
How about generate a new ConfigChangeEvent here? 

Subclass's ConfigChangeEvent  + resolved interestedChangedKeys  ==> new ConfigChangeEven

Reply:
How about we create a subclass of `ConfigChangeEvent`, which contains the original config change event and the interested keys so that we don't need to create a new instance?

Reply:
When subclasses of `AbstractConfig` invoke method `com.ctrip.framework.apollo.internals.AbstractConfig#fireConfigChange`, they don't know about any information of `interestedKeys` because `interestedKeys` are saved in `AbstractConfig`.

If we create a subclass of `ConfigChangeEvent`, when method `com.ctrip.framework.apollo.internals.AbstractConfig#fireConfigChange` is invoking, we still need to create a new instance of `subclass of ConfigChangeEvent`, because the listener need to get this instance?

Reply:
you are right, I was thinking we may use the `composition` pattern to avoid create a lot of `ConfigChangeEvent` instances.

Reply:
Change `ConfigChangeEvent` to an abstract class?

```java
public abstract class ConfigChangeEvent {
// content
}
```

---

Or create a new class `AbstractConfigChangeEvent`, 

```java
public abstract class AbstractConfigChangeEvent {
  public abstract Set<String> changedKeys();
  public abstract Set<String> interestedChangedKeys();
  public abstract ConfigChange getChange(String key);
  public abstract boolean isChanged(String key);
  public abstract String getNamespace();
}
```

then let `ConfigChangeEvent` extends it?

```java
public class ConfigChangeEvent extends AbstractConfigChangeEvent {
// ...
}
```

Reply:
My original idea was to create an abstract class or an interface, and then use `composition` to wrap the original `ConfigChangeEvent` into InterestedConfigChangeEvent. But the current implementation also looks good to me.